safrole/safrole.asn

```diff
@@ -2,114 +2,121 @@
+    ed25519 Ed25519Key,
     bls BlsKey,
-    metadata SEQUENCE (SIZE(128)) OF u8
+    metadata SEQUENCE (SIZE(128)) OF U8
 }
 
-ValidatorsData ::= SEQUENCE (SIZE(validators_count)) OF ValidatorData
+ValidatorsData ::= SEQUENCE (SIZE(validators-count)) OF ValidatorData
 
 TicketEnvelope ::= SEQUENCE {
-    attempt u8,
-    signature SEQUENCE (SIZE(784)) OF u8
+    attempt U8,
+    signature SEQUENCE (SIZE(784)) OF U8
 }
 
 EpochMark ::= SEQUENCE {
     entropy OpaqueHash,
-    validators SEQUENCE (SIZE(validators_count)) OF BandersnatchKey
+    validators SEQUENCE (SIZE(validators-count)) OF BandersnatchKey
 }
 
-TicketsMark ::= SEQUENCE (SIZE(epoch_length)) OF TicketBody
+TicketsMark ::= SEQUENCE (SIZE(epoch-length)) OF TicketBody
 
 -- Output markers.
 OutputMarks ::= SEQUENCE {
     -- New epoch signal.
-    epoch_mark [0] EpochMark OPTIONAL,
+    epoch-mark [0] EpochMark OPTIONAL,
     -- Tickets signal.
-    tickets_mark [1] TicketsMark OPTIONAL
+    tickets-mark [1] TicketsMark OPTIONAL
 }
 
 -- State relevant to Safrole protocol
 State ::= SEQUENCE {
-    -- tau: The most recent block's timeslog.
-    timeslot u32,
-    -- eta: The entropy accumulator and epochal randomness.
-    entropy SEQUENCE (SIZE(4)) OF OpaqueHash,
-    -- lambda: The validator keys and metadata which were active in the prior epoch.
-    prev_validators ValidatorsData,
-    -- kappa: The validator keys and metadata currently active.
-    curr_validators ValidatorsData,
-    -- gamma_k: The validator keys for the following epoch.
-    next_validators ValidatorsData,
-    -- iota: The validator keys and metadata to be drawn from next.
-    designed_validators ValidatorsData,
-    -- gamma_a: The sealing-key contest ticket accumulator.
-    tickets_accumulator SEQUENCE (SIZE(0..epoch_length)) OF TicketBody,
-    -- gamma_s: The sealing-key series of the current epoch.
-    tickets_or_keys TicketsOrKeys,
-    -- gamma_z: The Bandersnatch ring root.
-    tickets_verifier_key SEQUENCE (SIZE(384)) OF u8
+    -- Most recent block's timeslot.
+    tau U32,
+
+    -- Entropy accumulator and epochal randomness.
+    eta SEQUENCE (SIZE(4)) OF OpaqueHash,
+
+    -- Validator keys and metadata which were active in the prior epoch.
+    lambda ValidatorsData,
+
+    -- Validator keys and metadata currently active.
+    kappa ValidatorsData,
+
+    -- Validator keys for the following epoch.
+    gamma-k ValidatorsData,
+
+    -- Validator keys and metadata to be drawn from next.
+    iota ValidatorsData,
+
+    -- Sealing-key contest ticket accumulator.
+    gamma-a SEQUENCE (SIZE(0..epoch-length)) OF TicketBody,
+
+    -- Sealing-key series of the current epoch.
+    gamma-s TicketsOrKeys,
+
+    -- Bandersnatch ring commitment.
+    gamma-z SEQUENCE (SIZE(144)) OF U8
 }
 
 -- Input for Safrole protocol.
 Input ::= SEQUENCE {
     -- Current slot.
-    slot u32,
+    slot U32,
     -- Per block entropy (originated from block entropy source VRF).
     entropy OpaqueHash,
```

Comment:
Is this:  
- Blake2 of some VRF output
- VRF output
- VRF output hash?

I cannot parse it as compact VRF output, hence i assume it is the first. I know that it does not really matter for the sake of Safrole, but would be still good to know.

Reply:
These are 32 octets taken from the VRF output hash of the entropy source signature found in the header (H_v).

So in the "real" scenario, given the ietf-signature that you find in the header (H_v) you:
1) Deserialize it in a [`IetfVrfSignature`](https://github.com/davxy/bandersnatch-vrfs-spec/blob/a92814b468cf86bf17da6cab20321690ca9b545f/example/src/main.rs#L12) structure
2) `let entropy: [u8; 32] = signature.output.hash()[..32].try_into().unwrap();` (or something like that)
   (similar to [this](https://github.com/davxy/bandersnatch-vrfs-spec/blob/a92814b468cf86bf17da6cab20321690ca9b545f/example/src/main.rs#L196)
   
This is how the entropy is properly computed from the block header H_v component (GP expr 60).

As you mentioned, this is not the case in the test vectors as in this context the origin of the 32 octets is not really important. I guess we'll publish some blocks test vectors to verify / import. There you'll need this knowledge

Reply:
Thanks!